### ISSUE TITLE
Port twisted.python.threadpool to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5975]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5975|
|Type|enhancement|
|Created|2012-09-13 18:44:01Z|
|Branch|https://github.com/twisted/twisted/tree/threadpool-py3-5975-2|

`twisted.python.threadpool` is used by the reactor for the default DNS lookups implementation, and therefore needs to be ported.

<details><summary>Searchable metadata</summary>

```
trac-id__5975 5975
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_threadpool_py3_5975_2 branches/threadpool-py3-5975-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347561841000000 1347561841000000
changetime__1347632102000000 1347632102000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35636]) Branching to 'threadpool-py3-5975'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

Started buildbot run, haven't looked at it yet:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/threadpool-py3-5975

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

  1. There's a conflict in `_twistedpython3.py`.
  1. The `_waitForLock` change allocates an extra 30MB of memory on Python 2.  Can we do this loop without that extra cost?
  1. I bet some parts of `threadpool.py` would be clearer with context managers.  For example, in `_worker`, `with working(self): ...` instead of matched append/remove calls.  Would make a good ticket.

Uhh otherwise I guess this looks okay.  Please merge after addressing the above.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [35654]) Branching to 'threadpool-py3-5975-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35657]) Merge threadpool-py3-5975-2.

Author: itamar
Review: exarkun
Fixes: [#5975](https://github.com/twisted/twisted/issues/5975)

Port twisted.python.threadpool to Python 3.
